packages/modules/viewer-react/src/utilities/MakeCancellable.ts

```diff
@@ -0,0 +1,57 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+export function makeCancellable(
+  fn: (args?: any[]) => Generator<Promise<any> | Promise<void>, void, any>,
+  ...args: any[]
+): { promise: Promise<any>; cancel: () => void } {
+  const gen = fn(...args);
+  let cancelled = false;
+  let cancel: () => void = () => {
+    // nothing to cancel
+  };
+  const promise = new Promise((resolve, reject) => {
+    cancel = () => {
+      cancelled = true;
+      reject({ reason: "cancelled" });
+    };
+
+    onFulfilled();
+
+    function onFulfilled(res?: any) {
+      if (!cancelled) {
+        let result;
+        try {
+          result = gen.next(res);
```

Comment:
possibly misunderstanding but is this attempting to pass the response from the previous iteration into the next iterator? Is that right?

Reply:
yes, similar to how ```then``` methods work on promises. This whole thing can be thought of a giant promise chain with checking between statements